builder/azure/arm/step_capture_image.go

```diff
@@ -92,42 +102,58 @@ func (s *StepCaptureImage) Run(ctx context.Context, state multistep.StateBag) mu
-			} else if isSIGImage {
-				// It's possible to create SIG image
-				return multistep.ActionContinue
-			} else {
-				s.say("Capturing VHD ...")
-				err = s.captureVhd(ctx, vmId, vmCaptureParameters)
-			}
-		}
 		if err != nil {
 			state.Put(constants.Error, err)
 			s.error(err)
 
 			return multistep.ActionHalt
 		}
-
-		// HACK(chrboum): I do not like this.  The capture method should be returning this value
-		// instead having to pass in another lambda.
-		//
-		// Having to resort to capturing the template via an inspector is hack, and once I can
-		// resolve that I can cleanup this code too.  See the comments in azure_client.go for more
-		// details.
-		// [paulmey]: autorest.Future now has access to the last http.Response, but I'm not sure if
-		// the body is still accessible.
-		template := s.get(s.client)
-		state.Put(constants.ArmCaptureTemplate, template)
+		if isManagedImage {
+			s.say("Capturing image ...")
+			var targetManagedImageResourceGroupName = state.Get(constants.ArmManagedImageResourceGroupName).(string)
+			var targetManagedImageName = state.Get(constants.ArmManagedImageName).(string)
+			var targetManagedImageLocation = state.Get(constants.ArmLocation).(string)
+			s.say(fmt.Sprintf(" -> Image ResourceGroupName   : '%s'", targetManagedImageResourceGroupName))
+			s.say(fmt.Sprintf(" -> Image Name                : '%s'", targetManagedImageName))
+			s.say(fmt.Sprintf(" -> Image Location            : '%s'", targetManagedImageLocation))
+			err = s.captureManagedImage(ctx, subscriptionId, targetManagedImageResourceGroupName, targetManagedImageName, imageParameters)
+			if err != nil {
+				state.Put(constants.Error, err)
+				s.error(err)
+
+				return multistep.ActionHalt
+			}
+		} else if isSIGImage {
+			// It's possible to create SIG image without a managed image
+			return multistep.ActionContinue
+		} else {
+			// VHD Builds are created with a field called the VMId in its name
+			// Get that ID before capturing the VM so that we know where the resultant VHD is stored
+			vmInternalID, err := s.getVMInternalID(ctx, vmId)
+			if err != nil {
+				err = fmt.Errorf("Failed to get build VM before capturing image with err : %s", err)
+				state.Put(constants.Error, err)
+				s.error(err)
+				return multistep.ActionHalt
+			} else {
+				if vmInternalID == "" {
+					err = errors.New("Failed to get build VM before capturing image, Azure did not return the field VirtualMachine.Properties.VMId")
+					state.Put(constants.Error, err)
+					s.error(err)
+					return multistep.ActionHalt
+				} else {
+					s.say(fmt.Sprintf(" -> VM Internal ID            : '%s'", vmInternalID))
+					state.Put(constants.ArmBuildVMInternalId, vmInternalID)
+					s.say("Capturing VHD ...")
+					err = s.captureVhd(ctx, vmId, vmCaptureParameters)
+					if err != nil {
+						state.Put(constants.Error, err)
```

Comment:
This code is a bit messy now but I was having errors where the error was not correctly being passed into nested loops, so this felt simpler